### PR TITLE
cli: branch: on create/set/rename, assume deleted tracking local branch still exists

### PR DESCRIPTION
**Pull request by yuja**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: cli/src/commands/branch.rs [hunk 1]
```diff
@@ -251,14 +251,19 @@ fn cmd_branch_create(
...
+        if view.get_local_branch(name).is_present() {
+            return Err(user_error_with_hint(
+                format!("Branch already exists: {name}"),
+                "Use `jj branch set` to update it.",
+            ));
+        }
+        if has_tracked_remote_branches(view, name) {
+            return Err(user_error_with_hint(
+                format!("Tracked remote branches exist for deleted branch: {name}"),
+                "Use `jj branch set` to restore local branch.",
```
Comment (ilyagr) [inline cli/src/commands/branch.rs:264]: Optional:

I think it'd be nicer if this printed `name@remote`, not just `name`.

Also, I think you could add "..., or `jj git push` to delete the branch on remote(s) before trying again" (assuming this is true).

Comment (yuja) [reply cli/src/commands/branch.rs:264]: > I think it'd be nicer if this printed `name@remote`, not just `name`, if it's easy. Or it could say "... on remotes A, B, C".

I tend to agree, but I wanted to keep the implementation simple. It's unlikely that user has to deal with this situation often.

> Also, I think you could add "..., or `jj git push` to delete the branch on remote(s) before trying again" to the hint (assuming this is true).

Hmm, suppose the user doesn't remember why he deleted tracking local branch, `jj branch forget` (or `untrack`) is a better option. WDYT?

Comment (ilyagr) [reply cli/src/commands/branch.rs:264]: The best I could come up with so far is:

> Use `jj branch set` to restore the local branch, `jj branch untrack 'glob:{name}@*'` to forget that this branch existed locally.

I think both the behavior of `jj branch forget` in this situation and the reason that the above instruction works are a bit confusing. I wonder whether we should have `jj forget --local {name}` be an alias for `jj branch untrack 'glob:{name}@*'`. This makes sense especially if we think that the behavior of the two might diverge later.

I'm also wondering whether the first clause should say:

> Use `jj branch s {name} -r {name}@REMOTE`...

File: cli/src/commands/branch.rs [hunk 2]
```diff
@@ -399,6 +393,13 @@ fn cmd_branch_set(
     Ok(())
 }
 
+/// Whether or not the `branch` has any tracked remotes (i.e. is a tracking
+/// local branch.)
+fn has_tracked_remote_branches(view: &View, branch: &str) -> bool {
```
Comment (ilyagr) [inline cli/src/commands/branch.rs:412]: (Optional) Re making the code more difficult from https://github.com/martinvonz/jj/pull/3884/commits/88ff3555ff40e8e70e51fdd5932ce97cbdadbcdd#r1640243864: I don't know that it would be that much more difficult to return an iterator or a vector here, but maybe I'm missing something.

In any case, adding the remote names to the message (or not) is non-blocking for the PR.

File: cli/src/commands/branch.rs [hunk 3]
```diff
@@ -302,6 +307,11 @@ fn cmd_branch_rename(
     if view.get_local_branch(new_branch).is_present() {
         return Err(user_error(format!("Branch already exists: {new_branch}")));
     }
+    if has_tracked_remote_branches(view, new_branch) {
+        return Err(user_error(format!(
+            "Tracked remote branches exist for deleted branch: {new_branch}"
```
Comment (ilyagr) [inline cli/src/commands/branch.rs:312]: If we decide to add something about `untrack` or `branch forget` in `jj branch delete`, we should probably add a hint about that here as well.

**Review by ilyagr — APPROVED** [review]
I am still undecided on the best possible message, but it's OK already, and that can be fixed up later. The rest of the PR LGTM.

Thank you!

File: cli/src/commands/branch.rs (hunk 1 above)
Comment (yuja) [reply cli/src/commands/branch.rs:264]: > The best I could come up with so far is:
> 
> > Use `jj branch set` to restore the local branch, `jj branch untrack 'glob:{name}@*'` to forget that this branch existed locally.

Sounds good, thanks. I would rephrase the latter part as "to untrack remote branches." `"forget"` in this context implies remote-tracking branches will be deleted locally.

> I think both the behavior of `jj branch forget` in this situation and the reason that the above instruction works are a bit confusing. I wonder whether we should have `jj forget --local {name}` be an alias for `jj branch untrack 'glob:{name}@*'`. This makes sense especially if we think that the behavior of the two might diverge later.

Perhaps, it's the matter whether `forget NAME` should delete untracked remotes or not? I don't think the current `jj forget` behavior is correct, but it generally works as expected.

BTW, I don't expect `forget --local {name}` is equivalent to `untrack 'glob:{name}@*'`. The former should delete remote-tracking branches, but the latter shouldn't.

> I'm also wondering whether the first clause should say:
> 
> > Use `jj branch s {name} -r {name}@REMOTE`...

It seems too specific. The user's intent is unclear here. Maybe "to restore the local branch" in the hint is also too specific.

File: cli/src/commands/branch.rs (hunk 2 above)
Comment (yuja) [reply cli/src/commands/branch.rs:412]: It can return a `Vec`. Returning an iterator has lifetime issue.

That said, the things I wasn't bothered to are to provide different hints depending on the remote name (`@git` vs the others), and writing tests for various combinations.

**Review by yuja — COMMENTED** [review]
FWIW, this PR can go away if we decide to reorganize `create` + `set` as something like `set` (upsert) + `move`.

https://github.com/martinvonz/jj/issues/3584

The weirdness I'm trying to fix by this patch is that "create" preserves the existing tracking state. If the command name were "set", the behavior makes sense.

**Comment by yuja** [comment]
> FWIW, this PR can go away if we decide to reorganize `create` + `set` as something like `set` (upsert) + `move`.

Created https://github.com/martinvonz/jj/pull/3895

I personally prefer it over this PR since it seems more useful and will have a fewer special cases.

**Comment by ilyagr** [comment]
Wouldn't this problem affect `jj branch set --new` just as much as `jj branch create`?

**Comment by yuja** [comment]
> Wouldn't this problem affect `jj branch set --new` just as much as `jj branch create`?

My feeling is that `set --new` is okay to preserve existing tracking state (and maybe show warning?) `--new` might be a bad name, but it's the flag to not move existing branches. I feel "create" has stronger guarantee.

**Comment by ilyagr** [comment]
I don't think I would agree if `jj branch set --create` *only* created branches.

If `jj branch set --create` creates-or-moves (aka upserts, I think) branches, this makes more sense. I'm guessing this is your preference.

Some less solidified thoughts:

This might create a (minor? not minor?) problem if the user deleted a branch `mybranch`, thought it was gone, and then did `jj branch set --create mybranch` with the intention of creating a new branch. Instead, they would get a branch that's tracked by random remote branches they didn't know about. So, this seems like a (minor?) problem with the whole upsert idea that would be avoided by having `jj branch create` (possibly in addition to `jj branch set --create`).

I am not sure this is a fatal problem. The same problem would also happen if you `jj branch set --create`-d a new `mybranch` before there were any remote-tracking branches, and then fetched from a remote that has `mybranch` on it. So, maybe it's OK to have the same problem in the `jj branch set --create mybranch` case, I'm unsure.

Finally, if `jj branch set --create` upserts branches, I am realizing I do have a mild preference for keeping `jj branch create` for one more reason. As I was writing last paragraph, it was much more natural to write "if you `jj branch create mybranch`" as opposed to "if you `jj branch set --create`-d a new `mybranch`".

If we keep `jj branch create` in any form, I do think something like this PR would be useful.

**Comment by ilyagr** [comment]
As an aside, I'm pretty sure I want `jj branch forget --local` to exist. It's not actually exactly the same as `jj branch untrack glob:branch@*` if the branch exists. Once it exists, we could put it into the hints of this PR (only if we continue with this PR, of course, I was mostly thinking about it before I saw the other one).

Most likely, I'll require people to specify either `--local` or `--global` (and perhaps `--from-remotes` that lists specific remotes or a glob would be a third option eventually). In the long term, I think I prefer either `--local` being the default or there being no default.

Let me know if you prefer not doing this or find this controversial enough to benefit from filing an FR before trying to implement a PR.

File: cli/src/commands/branch.rs (hunk 1 above)
Comment (ilyagr) [reply cli/src/commands/branch.rs:264]: I read this late, but I sort-of answered this in https://github.com/martinvonz/jj/pull/3884#issuecomment-2172040123.

I think the current `forget` is awkward and shouldn't be the default. I currently think about calling it `forget --global`, and requiring the user to specify either `--local` or `--global`. In the long term, I'd probably be OK with `--local` becoming the default, but I'm not sure that's necessary.

**Comment by yuja** [comment]
> I don't think I would agree if `jj branch set --create` _only_ created branches.
> 
> If `jj branch set --create` creates-or-moves (aka upserts, I think) branches, this makes more sense. I'm guessing this is your preference.

My plan is:
* make `jj branch set` do upsert by default (i.e. create-or-move)
* add `--new` (or deny-move) flag for users who want to ensure that the branch doesn't exist.

The reason behind the idea is that `jj config set` does upsert, and if we add `jj tag set`, I think it would do upsert with `--allow-overwrite` flag or something. (It wouldn't make sense to add `tag set` for very narrow use case, in addition to `tag create`.)

> This might create a (minor? not minor?) problem if the user deleted a branch `mybranch`, thought it was gone, and then did `jj branch set --create mybranch` with the intention of creating a new branch. Instead, they would get a branch that's tracked by random remote branches they didn't know about.

Yeah. That's why I think we might want some warning message.

Another reason I'm getting skeptical about this PR is that `jj branch rename old new && jj branch rename new old` will no longer work if `old` had tracking remotes and renaming back to deleted tracking branch is a hard error.

File: cli/src/commands/branch.rs (hunk 1 above)
Comment (yuja) [reply cli/src/commands/branch.rs]: Update hint to suggest both `jj branch set` and `jj branch untrack glob:..`. PTAL.

File: cli/src/commands/branch/rename.rs [hunk 4]
```diff
@@ -60,11 +60,11 @@ pub fn cmd_branch_rename(
     if has_tracked_remote_branches(view, old_branch) {
         writeln!(
             ui.warning_default(),
-            "Branch {old_branch} has tracking remote branches which were not renamed."
+            "Branch {old_branch} has tracked remote branches which were not renamed."
```
Comment (ilyagr) [inline cli/src/commands/branch/rename.rs:63]: (Optional): I would prefer "is tracking remote branches" to "has tracked", since the latter sounds like past tense.

Another option (my second choice) would be "Tracking remote branches for branch {old_branch} were not renamed".

Thanks for fixing this, BTW!

File: cli/tests/test_branch_command.rs [hunk 5]
```diff
@@ -354,8 +354,8 @@ fn test_branch_rename() {
     let (_stdout, stderr) =
         test_env.jj_cmd_ok(&repo_path, &["branch", "rename", "bremote", "bremote2"]);
     insta::assert_snapshot!(stderr, @r###"
-    Warning: Branch bremote has tracking remote branches which were not renamed.
-    Hint: to rename the branch on the remote, you can `jj git push --branch bremote` first (to delete it on the remote), and then `jj git push --branch bremote2`. `jj git push --all` would also be sufficient.
+    Warning: Branch bremote has tracked remote branches which were not renamed.
```
Comment (ilyagr) [inline cli/tests/test_branch_command.rs:399]: Same as above.

**Review by ilyagr — APPROVED** [review]
Looks great, thank you!

File: cli/src/commands/branch/create.rs [hunk 6]
```diff
@@ -42,14 +43,22 @@ pub fn cmd_branch_create(
...
+            return Err(user_error_with_hint(
+                format!("Branch already exists: {name}"),
+                "Use `jj branch set` to update it.",
+            ));
+        }
+        if has_tracked_remote_branches(view, name) {
+            return Err(user_error_with_hint(
+                format!("Tracked remote branches exist for deleted branch: {name}"),
+                format!(
+                    "Use `jj branch set` to reset local branch. Run `jj branch untrack \
```
Comment (ilyagr) [inline cli/src/commands/branch/create.rs:57]: "reset" is a bit confusing here, let's do "recreate"?

File: cli/src/commands/branch/set.rs [hunk 7]
```diff
@@ -50,7 +50,7 @@ pub fn cmd_branch_set(
     let mut new_branch_names: Vec<&str> = Vec::new();
     for name in branch_names {
         let old_target = repo.view().get_local_branch(name);
-        if old_target.is_absent() {
+        if old_target.is_absent() && !has_tracked_remote_branches(repo.view(), name) {
```
Comment (ilyagr) [inline cli/src/commands/branch/set.rs:53]: Perhaps copy the `// Restoring local target shouldn't invalidate tracking state` comment here?

**Review by ilyagr — APPROVED** [review]

File: cli/src/commands/branch/set.rs (hunk 7 above)
Comment (ilyagr) [reply cli/src/commands/branch/set.rs:53]: Or maybe it could say something like

> If the branch is missing and tracking remote branches, we are resurrecting the local branch and we shouldn't change it's tracking state.

File: cli/src/commands/branch/create.rs (hunk 6 above)
Comment (yuja) [reply cli/src/commands/branch/create.rs]: It's not "creating" new branch (in a sense of `jj branch create`), but seems okay here. Updated.

File: cli/src/commands/branch/rename.rs (hunk 4 above)
Comment (yuja) [reply cli/src/commands/branch/rename.rs]: Good point. I choose the latter. The former could be parsed as "[Branch] is [tracking remote branches]".

File: cli/src/commands/branch/set.rs (hunk 7 above)
Comment (yuja) [reply cli/src/commands/branch/set.rs:55]: Added comment (but without "we shouldn't change it's tracking state.") Tracking state is unrelated in this context. `new_branch_names` is used only for emitting a warning.

**Review by yuja — COMMENTED** [review]
Thanks for suggestions.

File: cli/src/commands/branch/rename.rs (hunk 4 above)
Comment (ilyagr) [reply cli/src/commands/branch/rename.rs]: Ah, thanks for finding and fixing the typo in my suggestion. What you merged is correct, I think.
